### DISCUSSION TITLE
SRP Client Lease Interval and Key Lease Interval

### DISCUSSION THREAD

**Discussion by OlivierGre** [openthread/openthread#7955] [Q&A] [answered]
After a Thread commissioning, I'm using SRP to advertize the presence of the Thread device. It is working fine.
Can you please explain me the respective roles of Lease Interval and Key Lease Interval ? What will happen at the end of this durations?

Is it possible to use them to stop the SRP service after X seconds?

Thank you

**Comment by jwhui** [comment, marked as answer]
The lease specifies how long the DNS-SD PTR, SRV, A, AAAA, and TXT records will remain valid. This helps ensure that the service information is no longer discoverable within a reasonable amount of time after the device publishing the information disconnects from the network.

The key lease specifies how long the KEY record will remain valid, which allows the device to retain its claim on the name for much longer than the lease interval even when it is disconnected from the network.

Note that the SRP registrar can override the lease and key lease values requested by the SRP client. See [otSrpServerSetLeaseConfig()](https://openthread.io/reference/group/api-srp#otsrpserversetleaseconfig) to configure the bounds on the SRP registrar.

You can find more information about the lease and key lease in [draft-ietf-dnssd-srp](https://datatracker.ietf.org/doc/html/draft-ietf-dnssd-srp).

**Reply by OlivierGre** [reply]
OK thank you Jonathan. 
In the future, I will put my questions directly in "Discussions". In fact I had never noticed its presence. Sorry for that.


> The lease specifies how long the DNS-SD PTR, SRV, A, AAAA, and TXT records will remain valid. This helps ensure that the service information is no longer discoverable within a reasonable amount of time after the device publishing the information disconnects from the network.

Ok good, this is what I need. When the device gets disconnected from Thread network, I want the SRP service to stop pretty quickly (let's say after 60 seconds). 
I can see that I should change SRP server config on my RCP to adjust the min bound ("min lease" value). For the moment I have:
```
> srp server lease
min lease: 1800
max lease: 7200
min key-lease: 86400
max key-lease: 1209600
```

> The key lease specifies how long the KEY record will remain valid, which allows the device to retain its claim on the name for much longer than the lease interval even when it is disconnected from the network.

I think that I will not have to change this key lease interval. I can see that I should take care if I have 2 thread devices. I should take care that they don't try to create the same SRP service otherwise this claiming on the name will maybe cause some problems.
One solution seems to include the device EUI in one of the names: Shall I put it in the host name, instance name or service name? My bet would be to  append it to the instance name. 

Thank you
Regards
Olivier


> 
> Note that the SRP registrar can override the lease and key lease values requested by the SRP client. See [otSrpServerSetLeaseConfig()](https://openthread.io/reference/group/api-srp#otsrpserversetleaseconfig) to configure the bounds on the SRP registrar.
> 
> You can find more information about the lease and key lease in [draft-ietf-dnssd-srp](https://datatracker.ietf.org/doc/html/draft-ietf-dnssd-srp).

**Reply by wgtdkp** [reply]
To avoid name conflicts, you should try to use unique names for both the host name and service instance name - any name conflicts in the DNS RRs will result in a failure of SRP registration and all SRP services & host in the SRP message will be rejected.

**Reply by OlivierGre** [reply]
Ok thanks.

If the device's EUI is "0123456789ABCDEF" I would then use:

```
#define SRP_CLIENT_HOST_NAME   "ot-host-0123456789ABCDEF"
#define SRP_SERVICE_INSTANCE_NAME      "ot-service-0123456789ABCDEF"
#define SRP_SERVICE_NAME       "_ipps._tcp"
```

**Reply by abtink** [reply]
Couple of things to point out which I hope may help:
- The client will try to refresh the service registration (send a new SRP update to registrar/server) when the granted lease is about to expire. 
- So using a shorter lease interval will increase traffic (more frequent SRP Update messages from the clients).
- The OT `Srp::Client` implementation uses a guard time to renew a bit earlier than the granted lease expiration time.
- We have a build-time OT config param which determines this and the default value is set to 120 sec (2 minutes).
```c++
/**
 * @def OPENTHREAD_CONFIG_SRP_CLIENT_LEASE_RENEW_GUARD_INTERVAL
 *
 * Specifies the guard interval (in seconds) for lease renew time. The guard interval determines how much earlier
 * (relative to the lease expiration time) the SRP client will send an SRP update for lease renewal.
 *
 */
#ifndef OPENTHREAD_CONFIG_SRP_CLIENT_LEASE_RENEW_GUARD_INTERVAL
#define OPENTHREAD_CONFIG_SRP_CLIENT_LEASE_RENEW_GUARD_INTERVAL 120 // two minutes in seconds
#endif
```
- In general, while it is technically possible to use/allow shorter lease intervals (~60sec), it may not be the primary intended use-case/behavior (may require some tuning of parameters).

**Reply by OlivierGre** [reply]
Thank you Abtin, that clearly helps.

I can tell you more about what I'm doing:
My demo is about NFC Commissioning:
I have an OTBR, an Android phone acting as the Commissioner (running our own app) and a Thread device that I want to add to the Thread network.
The Thread device has an NFC tag. This is not a basic tag but a tag connected to the MCU via I2C. The tag contains the same information as what you put in a QRCode.
To initiate the commissioning, the user put the smartphone on the NFC Tag. The smartphone reads the tag content and gets the EUI64 and the PSKd.  NFC communication is also used at that time to initiate the otJoinerStart() action on the Thread device.
Once the commissioning done, the Thread device uses SRP to advertize its IP address. The smartphone scan "_ipps._tcp" services, finds the one with the expected EUI64 ("ot-host-0123456789ABCDEF", "ot-service-0123456789ABCDEF") and can then communicate with it. For the moment I use UDP.

Our goal is to show this commissioning so when this is done, we reset the Thread device. Credentials have not been kept in non volatile memory so it starts disconnected from Thread network and we can do the demo again. 

My problem is with SRP service peristance. If I try a new commissioning, previous SRP service is still there. This is sometimes a problem because the Thread device may receive a new IPV6 address. Once the commissioning done, SRP service will get updated with the new IPV6 address but it is too late because the smartphone has already seen the old IP address.

That's why I wanted a SRP service with a short duration.

I'm going to try the following strategy: Once the commissining done and the device discovery (via SRP) done, the smartphone and the Thread device will echange a message. The Thread device will then call otSrpClientRemoveHostAndServices() (with removeKeyLease = true and sendUnregToServer = true) so that the SRP Service get stopped.  

Your comments and suggestions are very welcome.
Thanks for your help
Regards

**Reply by jwhui** [reply]
If this is purely for demo purposes, I would think that setting the lease values to something small would be fine. While that would cause the device to refresh its lease more frequently, that should not be much of an issue for a demo.

**Reply by abtink** [reply]
> I'm going to try the following strategy: Once the commissining done and the device discovery (via SRP) done, the smartphone and the Thread device will echange a message. The Thread device will then call otSrpClientRemoveHostAndServices() (with removeKeyLease = true and sendUnregToServer = true) so that the SRP Service get stopped.

Yes. I think this can work. You can register a callback in SRP client to get notified on changes (so to tell when remove was done and updated on server) and then continue to next step (reset the device, etc).

**Reply by jwhui** [reply]
> > I'm going to try the following strategy: Once the commissining done and the device discovery (via SRP) done, the smartphone and the Thread device will echange a message. The Thread device will then call otSrpClientRemoveHostAndServices() (with removeKeyLease = true and sendUnregToServer = true) so that the SRP Service get stopped.
> 
> Yes. I think this can work. You can register a callback in SRP client to get notified on changes (so to tell when remove was done and updated on server) and then continue to next step (reset the device, etc).

The only downside with this approach I see is if that message does not get through for whatever reason (maybe because the device was factory-reset too early), then the symptom will be unaddressed.

**Reply by OlivierGre** [reply]
Ok thank you @jwhui and @abtink for your messages.

Is it possible to stop a SRP service that has not been started during the current run? I mean: Let's say that the Thread device has started a SRP service and then been rebooted without stopping the service. Can I stop the SRP service afterward? Is it possible to use otSrpClientRemoveHostAndServices() to stop a service not started during the current session?

**Comment by OlivierGre** [comment]
I'm thinking to another solution:
I can manage to get the host name and the service instance name to be different every times. 
The host name would still contain the EUI64 (ex: "ot-host-0123456789ABCDEF") but the service instance name would contain a random number known by the Thread device and by the Android phone (ex: "ot-service-AABBCCDD"). This random number would change every times so the smartphone will specifically search this service instance name and will ignore services with other number (remaining from previous demos).

By this way, I don't have to bother anymore with SRP service persistance. 
What do you think?

**Reply by jwhui** [reply]
SRP uses key records to claim a name. The device will randomly generate a new private key each time it is factory reset. As a result, the device will not be able to register the same name because it will have a different key.

**Reply by OlivierGre** [reply]
It is not clear to me which names should be unique: 
I was thinking that the couple (host name, service instance name) should be unique.

As per your message, I'm wondering if it is a problem to use the same host name but a different service instance name.

With what I have proposed above, I would get something like that:

Commissioning 1:
Host name: "ot-host-0123456789ABCDEF"
service instance name: "ot-service-AAAAAAAA"

Commissioning 2:
Host name: "ot-host-0123456789ABCDEF"
service instance name: "ot-service-BBBBBBBB"

Is this allowed?

**Reply by jwhui** [reply]
All names are protected with the same key record. So you would need to use a different name with a different key (which occurs after a factory reset).

**Reply by OlivierGre** [reply]
ok, thank you

**Reply by OlivierGre** [reply]
The discussion is now difficult to follow because of the nested messages. So I'm copying my last message here:

Ok thank you @jwhui and @abtink for your messages.

Is it possible to stop a SRP service that has not been started during the current run? I mean: Let's say that the Thread device has started a SRP service and then been rebooted without stopping the service. Can I stop the SRP service afterward? Is it possible to use otSrpClientRemoveHostAndServices() to stop a service not started during the current session?

**Reply by jwhui** [reply]
Yes, the documentation for [`otSrpClientRemoveHostAndServices()`](https://openthread.io/reference/group/api-srp#otsrpclientremovehostandservices) states:

> The `aSendUnregToServer` determines the behavior when the host info is not yet registered with the server. If `aSendUnregToServer` is set to `false` (which is the default/expected value) then the SRP client will immediately remove the host info and services without sending an update message to server (no need to update the server if nothing is yet registered with it). If `aSendUnregToServer` is set to `true` then the SRP client will send an update message to the server. Note that if the host info is registered then the value of `aSendUnregToServer` does not matter and the SRP client will always send an update message to server requesting removal of all info.
> 
> One situation where `aSendUnregToServer` can be useful is on a device reset/reboot, caller may want to remove any previously registered services with the server. In this case, caller can [otSrpClientSetHostName()](https://openthread.io/reference/group/api-srp#group__api-srp_1gad00aa82af3a71e91ef2d8eb796099a64) and then request [otSrpClientRemoveHostAndServices()](https://openthread.io/reference/group/api-srp#group__api-srp_1ga43324d90582bbb2f97b5bbaf31e97fe4) with aSendUnregToServer as true.

**Reply by OlivierGre** [reply]
Thank you!
